feature(ajax): Adds a new elgg/ajax module with unified API

mrclay opened the pull request:
All `elgg/Ajax` methods transparently handle server-side messages/errors, as well as return the raw value (the JSON wrapper is not exposed to the user as in `elgg.action()`.)

A client plugin hook allows modifying the request data, and both client and server-side hooks can modify the returned data.

You can test this by enabling the developers plugin and entering this at the JavaScript console: `require(["developers/ajax_demo"])`

The ajax views system can now serve static views that weren't specifically registered for ajax. This is mainly for fetching static `.html` view files.

Fixes #8323

(Basic differences from the 1.8 APIs: AMD, less guesswork in API usage, allows plugin collaboration, no JSON wrappers exposed when using actions, system messages/errors always displayed on client.)
- [x] wait until I can manually merge in #9210

mrclay committed:
feature(ajax): Adds a new elgg/Ajax AMD module with unified API

All `elgg/Ajax` methods transparently handle server-side messages/errors, as
well as return the raw value (the JSON wrapper is not exposed to the user
as in `elgg.action()`.)

A client plugin hook allows modifying the request data, and both client
and server-side hooks can modify the returned data.

You can test this by enabling the developers plugin and entering this at the
JavaScript console: `require(["developers/ajax_demo"])`

The ajax views system can now serve static views that weren't specifically
registered for ajax. This is mainly for fetching static `.html` view files.

Fixes #8323

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,221 @@
+========
 
-    // in myplugin/actions/do_math.php
+All perform the following actions:
```
All what?

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,221 @@
-        register_error('Sorry, Ajax only!');
-        forward();
-    }
+#. Client-side, the ``data`` option is filtered by the hook ``ajax_request_data, *``.
```
Perhaps `action:<action_name>` for clarity. Asterisk is a bit confusing.

hypeJunction commented on docs/guides/hooks-list.rst:
```diff
@@ -186,6 +182,30 @@ Action hooks
+
+**ajax_response, \***
+	When the ``elgg/ajax`` AMD module is used, this hook gives access to the response object
+	(``\Elgg\Services\AjaxResponse``) so it can altered/extended. The hook type depends on
```
can be

hypeJunction commented on engine/classes/Elgg/ActionsService.php:
```diff
@@ -329,57 +330,61 @@ public function exists($action) {
+
+		// always pass the full structure to avoid boilerplate JS code.
+		$params = array_merge($params, array(
+			'output' => '',
```
Should we perhaps send an empty object `'{}'` if `HTTP_ACCEPT` is `application/json`?

mrclay replied:
This is the 1.8 API. Maybe make a case for that in a separate issue?

hypeJunction commented on engine/classes/Elgg/Ajax/Service.php:
```diff
@@ -0,0 +1,200 @@
+	 * @return bool
+	 */
+	public function isAjax2Request() {
+		$version = _elgg_services()->request->headers->get('X-Elgg');
```
We might want to check the HMAC headers used in web services to avoid possible conflicts.

hypeJunction commented on views/default/elgg/ajax.js:
```diff
@@ -0,0 +1,199 @@
+	var elgg = require('elgg');
+	var spinner = require('elgg/spinner');
+
+	var spinner_enabled = true;
```
Could we make it a global config option? Then spinner can only be required if it's enabled.

mrclay replied:
Every user of the API would still need to set it, just in case another user set it differently. The solution is to put the methods on an instance created locally.

```js
var ajax = new require('elgg/Ajax');
ajax.useSpinner = false;
...
```

mrclay replied:
(I'd need to make it a constructor to do that)

hypeJunction commented:
First impression: great stuff!

juho-jaakkola commented:
Wow, there's a lot of stuff here! I'll need to be much less tired to dig through this... :)

mrclay commented:
I converted the AMD module to a constructor (eliminating the global spinner property), and addressed the  comments.

mrclay commented:
I've reworked the demo part of this to test the new client/server hooks and moved it into the developers plugin since it might be useful for learning. I'd really like to get this in so we can start standardizing core stuff on it.

As the commit says, you can test via console: `require(["developers/ajax_demo"])`

hypeJunction commented:
First comment, it feels we need a better way of referencing client and server side variables and methods in docs. E.g. it's hard to know if `hook` refers to client side or server side hook. Does readthedocs allow icons of sorts?

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+Notes for actions:
+
+* All hooks have type ``action:<action_name>``. In this case, "action:do_math".
```
client side hooks

hypeJunction replied:
Can we use full hook structure `<name>,<type>` in docs?

mrclay replied:
There will be 2 client-side hooks and 1 server-side, all with different hook names, all with this type. I guess I can make all those explicit in this section to get the point across.

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+Notes for actions:
+
+* All hooks have type ``action:<action_name>``. In this case, "action:do_math".
+* CSRF tokens are added to the data.
```
request data

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+Performing actions
+------------------
+
+To execute an action, use ``performAction('action_name', options)``:
```
could we use `ajax.` as prefix and note that it refers to `elgg/Ajax` module everywhere in the docs?

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+* All hooks have type ``action:<action_name>``. In this case, "action:do_math".
+* CSRF tokens are added to the data.
+* The default method is POST.
```
Maybe use POST and GET as code blocks?

mrclay replied:
What do you mean?

hypeJunction replied:
`POST` instead of POST

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+Notes for paths:
+
+* All hooks have type ``path:<url_path>``. In this case, "path:<myplugin_time>".
```
Does this apply to all segments or just the identifier?

hypeJunction replied:
"path:myplugin_time" without brackets

mrclay replied:
Right now the hook type will have the complete path (all segments). I foresee good arguments for either way... I suppose we could leave out fetchPath from this API. It just seemed useful.

hypeJunction replied:
I definitely find it more useful than `fetchView`, as I avoid ajax views as much as can. I believe that anything loaded via ajax should have a fully qualifying URL and should be reachable by the browser should JS fail to hijack a click action.
Perhaps we should allow fetching a full path, but only trigger a hook on the page handler ID?

mrclay replied:
Yeah. Hook handlers can sniff the full URL if they want.

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+.. warning::
+
+	Ajax views must treat ``$vars`` as completely untrusted user data. You may want to use ``get_input()``
```
Really? I thought ajax page handler sanitizes request data.

mrclay replied:
It's HTML sanitized. I'll clarify this.

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+* All hooks have type ``form:<action_name>``. In this case, "form:myplugin/add".
+* ``output`` will be a string with the rendered view.
+* The request data are injected into ``$vars`` in your view.
```
Anything happening with sticky values?

mrclay replied:
It's just like the existing `/ajax/form` page handler. I'm not sure what sticky behavior we want to do automatically...

hypeJunction replied:
I never used it, and I don't I think I ever will. I use full URL path, e.g. `blog/edit/$guid`, to make sure all the page owner and context logic stays intact, and only output the form from the page handler if it's ajax request.

mrclay replied:
> I use full URL path, e.g. `blog/edit/$guid`, to make sure all the page owner and context logic stays intact, and only output the form

Sounds like a great idea. How could core build an API around that? Maybe a separate issue?

hypeJunction replied:
Definitely a separate issue, and not easy to implement as a standard API. There are many pitfalls, using `elgg_is_xhr()` is not always enough to decide which part of the page to output. I was ajaxifying profile loading and needed an entire `.elgg-page-body`, but some other plugin was only returning profile details on ajax requests, so I had to actually fool it by sending bad `X-Requested-With` headers.

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,220 @@
+
+.. note::
+
+	Request data are **not** passed into the ``input/form`` view! If you need to set ``$vars`` in the
```
This is really confusing.

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -156,42 +338,42 @@ Here's an example of fetching a view that returns a JSON-encoded array of times:
- * The GET params will be passed as ``$vars`` to *your* view, **not** the ``input/form`` view.
- * If you need to set ``$vars`` in the ``input/form`` view, you'll need to use the ``("view_vars", "input/form")``
-   plugin hook (this can't be done client-side).
+* The GET params will be passed as ``$vars`` to *your* view, **not** the ``input/form`` view.
```
Confusing

hypeJunction commented on docs/guides/hooks-list.rst:
```diff
@@ -186,6 +182,30 @@ Action hooks
 
+.. _guides/hooks-list#ajax:
+
+Ajax
```
I think we need to document client-side hooks in a new section. 

> This is *not* triggered by the ``elgg/Ajax`` module.
IMHO, this shouldn't be necessary, if we document hooks separately.

mrclay replied:
None of these are client-side hooks.

hypeJunction commented:
Does this fix  #9027?

hypeJunction commented on mod/developers/actions/developers/ajax_demo.php:
```diff
@@ -0,0 +1,29 @@
+
+	register_error('Error from ajax demo response hook');
+}
+elgg_register_plugin_hook_handler('ajax_response', 'action:developers/ajax_demo', 'developers_ajax_demo_alter');
```
I wish we could do this for non-ajax actions...

mrclay replied:
A hook after the action completes?

hypeJunction replied:
Yep, filtering the action output after the action completes. But that's another story and another ticket.

hypeJunction commented:
Does this mean we can't use existing actions with ajax requests? You mention in docs that we should't call `forward()`.

mrclay commented:
> Does this fix #9027?

No, but doesn't make it any harder to fix. We should fix that in 1.12.

hypeJunction commented:
> The existing 1.8 APIs shouldn't be affected at all. Before the responses are sent we branch off if we 
> get a special header from elgg/Ajax.

That won't make upgrading to new API easy though... We can just replace `elgg.action` with `fetchAction` without modifying the action code if it uses `forward()`?

mrclay commented:
I'll have to re-study why allowing forward() was going to be a problem...

hypeJunction commented on engine/lib/elgglib.php:
```diff
@@ -1538,7 +1538,16 @@ function _elgg_ajax_page_handler($segments) {
+		// cacheable views are always allowed
+		if (!array_key_exists($view, $allowed_views) && !_elgg_services()->views->isCacheableView($view)) {
+			if ($ajax_api->isReady()) {
+				$ajax_api->respondWithError("Ajax view '$view' was not registered");
```
This will be visible to the user? i18n?

mrclay replied:
This shouldn't happen if the plugin is code correctly.

mrclay commented:
For one thing we don't have a good way to expose it to the client. I supposed it could be returned under the key `__forward_url` but only if the response is an object. I'd rather treat forward() as return via for Ajax requests.

hypeJunction commented:
`forward()` is evil and overused. Actions should really be returning data that we can filter and do stuff with.

mrclay commented:
@hypeJunction I've updated the docs, PTAL

mrclay commented:
But if the use case is an action that can handle either ajax or a regular page request, seems like forward() is a necessity. Isn't that the pattern you're using?

mrclay commented:
Of course we can branch based on elgg_is_xhr()...

hypeJunction commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,233 @@
+
+.. note::
+
+	Request data are passed into the form view only. If you need to add inject parameters into the
```
Only the request data are passed to the requested form view (i.e. as a third parameter accepted by `elgg_view_form()`). If you need to pass attributes or parameters of the form element rendered by the ``input/form`` view (i.e. normally passed as a second parameter to ``elgg_view_form()``), use the server-side hook ``view_vars, input/form``.

hypeJunction commented:
Strawman:

```php
$action_result = require_once($action_file);
if (is_array($action_result) && isset($action_result['forward_url'])) {
   forward($action_result['forward_url']);
   // or do ajax magic
}
```

mrclay commented:
This handles forward() just fine, so I removed the warning and mentioned under Actions that the URL isn't returned to the client.

hypeJunction commented:
LGTM. I haven't done any actual testing, but examples look solid.

mrclay commented:
Yeah, ideally actions and "pages" could return some structured value that gets turned into the response. Much more testable. I know the Drupal folks want to get away from render arrays though, so maybe something more like an HTTP response. For another ticket...

mrclay commented on docs/guides/hooks-list.rst:
```diff
@@ -186,6 +182,30 @@ Action hooks
+**output, ajax**
+	This filters the JSON output wrapper returned to the legacy ajax API (``elgg.ajax``, ``elgg.action``, etc.).
+	Plugins can alter the output, forward URL, system messages, and errors. For the ``elgg/Ajax`` AMD module,
+	use the ``ajax_response`` hook documented above.
```
@hypeJunction hopefully I've made this clearer

hypeJunction replied:
We can improve later if people have questions. LGTM :v:

mrclay commented on docs/guides/javascript.rst:
```diff
@@ -466,11 +466,17 @@ Plugins can trigger their own hooks:
 Available hooks
 ^^^^^^^^^^^^^^^
 
```
@hypeJunction All the client-side hooks are documented separately here. In the future we'll want to expand the JS docs, but this seems OK for now.

mrclay commented on engine/classes/Elgg/ActionsService.php:
```diff
@@ -329,74 +330,81 @@ public function exists($action) {
+		$ajax_api = _elgg_services()->ajax;
+		if ($ajax_api->isReady()) {
+			if (isset($http_codes[$reason])) {
+				$ajax_api->respondWithError($http_codes[$reason], $reason);
```
@hypeJunction I've rebased merging in your HTTP commit. If the request was sent with the new API, this just sends the error header with message, no data.

hypeJunction replied:
Ok.

mrclay commented:
This is rebased and ready. @juho-jaakkola Any interest in at least running the demo on this? I'm eager to get this in.

juho-jaakkola commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,234 @@
-        register_error('Sorry, Ajax only!');
-        forward();
-    }
+All the ajax methods perform the following actions (not the hook types depend on the method):
```
The part inside the parenthesis is a bit confusing.

juho-jaakkola commented on docs/guides/ajax.rst:
```diff
@@ -1,31 +1,234 @@
 
-    $arg1 = (int)get_input('arg1');
-    $arg2 = (int)get_input('arg2');
+#. Client-side, the ``data`` option is filtered by the hook ``ajax_request_data``.
```
Perhaps `ajax_request_data, {action name}`? (Or `ajax_request_data, <action name>}`? Which type of brackets do we usually use?)

juho-jaakkola commented on mod/developers/views/default/developers/ajax_demo.js:
```diff
@@ -0,0 +1,70 @@
+define(function(require) {
+	var Ajax = require('elgg/Ajax');
```
Is there a reason `Ajax` starts with capital letter? AFAIK all other module names have only lowercase letters.

mrclay replied:
It's a class constructor.

juho-jaakkola commented on views/default/elgg/Ajax.js:
```diff
@@ -0,0 +1,205 @@
+		 * @param {Object} options jQuery.ajax options. See fetch()
+		 * @returns {jqXHR}
+		 */
+		this.performAction = function (action, options) {
```
Nitpicking, but I'd have slight preference on just `action()` instead of `performAction()`.

juho-jaakkola commented:
LGTM

mrclay commented on docs/guides/hooks-list.rst:
```diff
@@ -186,6 +182,30 @@ Action hooks
+	fetchPath()       path:<url_path>
+	fetchView()       view:<view_name>
+	fetchForm()       form:<action_name>
+	================  ====================
```
Maybe I should just name the methods like the hook types: action(), path(), view(), form() ...

juho-jaakkola replied:
+1
